### DISCUSSION TITLE
Difference between Clientside Properties and Serverside Properties?

### DISCUSSION THREAD

**Discussion by PicassoCT** [wixtoolset/issues#7697] [Questions]
Hello everyone,
i have inherited a WIX installation and was given the task to extend it with a optional dialog to enter a token. This is happening via a call to two custom actions - one to detect the Token, one to write it after a inputDlg completes. The Custom Action runs succesfully and the value is set, according to the msi log. 

```
	Zeile 1847: Property(N): INSTALLTOKENREQUEST = 0
	Zeile 1860: Property(N): SecureCustomProperties = INSTALLTOKENREQUEST;WIX_UPGRADE_DETECTED
	Zeile 2044: MSI (s) (78!4C) [10:44:03:144]: PROPERTY CHANGE: Modifying INSTALLTOKENREQUEST property. Its current value is '0'. Its new value: '1'.
	Zeile 2831: Property(S): INSTALLTOKENREQUEST = 1
	Zeile 2844: Property(S): SecureCustomProperties = INSTALLTOKENREQUEST;WIX_UPGRADE_DETECTED
	Zeile 2985: Property(C): INSTALLTOKENREQUEST = 0
	Zeile 2998: Property(C): SecureCustomProperties = INSTALLTOKENREQUEST;WIX_UPGRADE_DETECTED
```

However for the UI-decision making, the client property Property(C) is used, which stays unchanged. How do i acess the server side property  Property(S). I did not find much documentation on this except for some stackoverflow posts, so i would approve even more if i could get some official microsoft documentation. 

Regards 

PS: Here is the wxs file.

```

<?xml version="1.0" encoding="UTF-8"?>
<Wix xmlns="http://schemas.microsoft.com/wix/2006/wi"
     xmlns:ui ="http://schemas.microsoft.com/wix/UIExtension">
	<Product Id="*" Name="Client" Language="1033" Version="$(var.ProductVersion)" Manufacturer="$(var.Manufacturer)" UpgradeCode="$(var.UpgradeCode)">
		<Package InstallerVersion="200" Compressed="yes" InstallScope="perUser" />
		<UI>
			<UIRef Id="WixUI_InstallDir" />
			<Publish Dialog="WelcomeDlg"
				Control="Next"
				Event="NewDialog"
				Value="InstallDirDlg"
				Order="2">1
			</Publish>
			<Publish Dialog="InstallDirDlg"
				Control="Back"
				Event="NewDialog"
				Value="WelcomeDlg"
				Order="2">1
			</Publish>
			<Publish Dialog="ExitDialog"
			  Control="Finish"
			  Event="DoAction"
			  Value="LaunchApplication">WIXUI_EXITDIALOGOPTIONALCHECKBOX = 1
			</Publish>
		</UI>
		<WixVariable Id="WixUIBannerBmp" Value="Banner.bmp"/>
		<WixVariable Id="WixUIDialogBmp" Value="Dialog.bmp"/>
		<InstallExecuteSequence>
			<Custom Action="IsTokenDefinedCA" After="InstallFiles">NOT Installed</Custom>
			<Custom Action="LaunchApplication" OnExit="success">1 = IS_UPDATE</Custom>
		</InstallExecuteSequence>
		<CustomAction Id="LaunchApplication"
			BinaryKey="WixCA"
			DllEntry="WixShellExec"
			Impersonate="yes" />
		<CustomAction Id="IsTokenDefinedCA" BinaryKey="CustomActionBinary" DllEntry="IsTokenDefined" Return="check"/>
		<Property Id ="INSTALLDIR">
			<RegistrySearch Id="RegistrySearchInstallDir" Type="raw" Root="HKCU" Key="Software\[Manufacturer]\[ProductName]" Name="InstallDir" Win64="no"/>
		</Property>
		<Property Id="INSTALLTOKENREQUEST" Value="0" Secure="yes" />
		...
	</Product>

	<Fragment>
		...
	</Fragment>

	<!--Directory cleanup when uninstall-->
	<Fragment>
		...
	</Fragment>

	<Fragment>
		<UI Id="ExitDlgUI">
			<Dialog Id="OptionalExitDlg" Height="350" Width="493" Title="Client Setup">
				<Control Id="headerText" Type="Text" Height="50" Width="220" X="160" Y="25" TabSkip="no" Transparent="yes" Text="{\Headline12} Token found"/>
				<Control Id="InstallBtn" Type="PushButton" Height ="21" Width="70" X="327"  Y="322" Text="OK" >
					<Publish Event="EndDialog" Value="Return">1</Publish>
				</Control>
			</Dialog>
		</UI>
		<UI Id="InputDlgUI">	
			<Binary Id="Background" SourceFile="Dialog.bmp" />
			<TextStyle Id="Standard8" FaceName="Tahoma" Size="8" Red="0" Green="0" Blue="1"  />
			<TextStyle Id="Headline12" FaceName="Tahoma" Size="12" Red="0" Green="0" Blue="1"  />

			<Dialog Id="InputDlg" Height="350" Width="493" Title="Client Setup">

				<Control Id="InstallBtn" Type="PushButton"  Height ="21" Width="70" X="327"  Y="322" Text="OK" >
					<Publish Event="DoAction" Value="UserInputTokenCA">1 </Publish>
					<Publish Event="EndDialog" Value="Return">1</Publish>
				</Control>
				<Control Id="CancelBtn" Type="PushButton" Text="Abbrechen" Height ="21" Width="70" X="420" Y="322" Cancel ="yes">
					<Publish Event="EndDialog" Value="Exit" />
				</Control>
			</Dialog>
		</UI>
		<InstallUISequence>
			<Show Dialog="InputDlg" After="ExecuteAction">	<![CDATA[INSTALLTOKENREQUEST AND( INSTALLTOKENREQUEST="1")]]>	</Show>
			<Show Dialog="OptionalExitDlg" After="ExecuteAction"><![CDATA[INSTALLTOKENREQUEST AND (INSTALLTOKENREQUEST="0")]]></Show>
		</InstallUISequence>
		...
		<CustomAction Id="UserInputTokenCA" BinaryKey="CustomActionBinary" DllEntry="UserInputToken" />	
	</Fragment>
	
	<Fragment>
	...
	</Fragment>
</Wix>

```

**Comment by BMurri** [comment]
Windows Installer shares property values only one way (from client to server). There is no sharing back. The assumption is that most data gathering is done in the UI (aka client) and all changes/storage of binaries or data will be done in the server, never the UI.

UI can display data, but that is transitory, and accomplished via messages.

So, writing the token should be done from the `InstallExecuteSequence` only, and should NOT be a graphical application. If an application needs the token, it should retrieve it from wherever you store it in the OS.

**Reply by PicassoCT** [reply]
Hi BMurri,
i think i undestood the security concept behind it. I still try to understand 

> So, writing the token should be done from the InstallExecuteSequence only, and should NOT be a graphical application. If an application > > needs the token, it should retrieve it from wherever you store it in the OS.

Im all for collecting data in gui - but the token is entered by the user if needed and only if. The detection of the need for the token insertion is done in the  `InstallExecuteSequence `.  There must be some way to marshall a sequence depending on what is installed (server knowledge).  Can i communicate that as a sort of component already being installed?

**Reply by BMurri** [reply]
The need for the token cannot be determined in `InstallUISequence` at all? Does detection require elevation or some other condition?

**Reply by PicassoCT** [reply]
It needs to know wether a file on the disk contains some information?

**Reply by BMurri** [reply]
File on the disk before you start installation, or file on the disk that is potentially altered because of your installation?

If the file/information is present before installation, you should move your custom action to the UI sequence and be done with it.

**Reply by PicassoCT** [reply]
Is there a way to trigger a property re-evaluation?

**Reply by BMurri** [reply]
Depends on who's writing the property. Actions can be repeated, abet usually indirectly (they need a different name to run again in the same sequence at a different point in the same sequence, but you can make a custom action that only calls another action).

**Comment by PicassoCT** [comment]
I found a solution. Write and Consume Files as messages into the installation-folder during the custom actions. Then use these "Messages" to trigger the optional dialogs. Once you give up on the framework its actually pretty straight forward. All Messages need to be cleaned up by the initial action though.

```
	<Property Id="INSTALLTOKENREQUEST" Value ="0">
			<DirectorySearch  Id="AppDataSearch" Path="[INSTALLDIR]" Depth="0">
				<FileSearch Id="SearchForInstallTokenRequest" Name="InstallToken.request"/>				
			</DirectorySearch>
		</Property>
```

```
		<InstallUISequence>
			<Show Dialog="InputDlg" After="ExecuteAction"><![CDATA[INSTALLTOKENREQUEST <> "0"]]></Show>
		
		</InstallUISequence>
```

**Reply by PicassoCT** [reply]
So, turns out this aint a solution either, because the FileSearch triggers initially and then never again.

**Reply by BMurri** [reply]
What produces the InstallToken.request file? Could you write/call a custom action in the UI sequence after the installation completes on the server side that checks for that file's existence and sets your INSTALLTOKENREQUEST property?
